File: Week 4/solutions/D_bridges.py
from collections import deque

INF = 10 ** 9 # Helpful way to store infinity as a large number

# Input format: adj[i] is a list of the neighbors of node i
#               is_single[i][j] tells whether the edge from i to adj[i][j] is a single-lane bridge
#               adj[i].size() == is_single[i].size() for all i
# The input is 1-indexed
def bridges(n, adj, is_single):
    # In a later week we'll talk about generalized shortest paths problems
    # Dijkstra's algorithm solves this problem too

    dist = [INF] * (n + 1)
    dq = deque() # The trick to this problem is to use a deque instead of a queue
                    # If a vertex isn't a bridge, we can push it to the front of the deque instead
    
    dq.append(1)
    dist[1] = 0
    while len(dq) > 0:
        u = dq.popleft()
        for i in range(len(adj[u])):
            v = adj[u][i]
            new_dist = dist[u] + (1 if is_single[u][i] else 0)
            if new_dist < dist[v]:
                dist[v] = new_dist
                if new_dist == dist[u]:
                    dq.appendleft(v)
                else:
                    # It's fine if we push a vertex 2 times; the algorithm still takes O(n) time.
                    # Because we can show each vertex won't be pushed more than twice, 
                    # we don't need to do any processing to make sure we don't push the same vertices.
                    dq.append(v)
    return dist[n]


n, m = map(int, input().split())
adj = [[] for _ in range(n + 1)]
is_single = [[] for _ in range(n + 1)]
for _ in range(m):
    u, v, is_single_lane = map(int, input().split())
    adj[u].append(v)
    is_single[u].append(is_single_lane == 1)
    adj[v].append(u)
    is_single[v].append(is_single_lane == 1)

print(bridges(n, adj, is_single))

File: Week 4/solutions/D_bridges.java
import java.io.*;
import java.util.*;

class Main{
    // Make sure when you submit, the class name is Main
	
    // Input format: adj.get(i) is a list of all the neighbors of node i
    //               is_single.get(i).get(j) tells whether the edge from i to adj.get(i).get(j) is a single-lane bridge
    //               adj.get(i).size() == is_single.get(i).size() for all i
    // The input is 1-indexed
    static int bridges(int n, ArrayList<ArrayList<Integer>> adj, ArrayList<ArrayList<Boolean>> is_single){
        ArrayList<Integer> dist = new ArrayList<Integer>();
        while(dist.size() < n + 1) dist.add(Integer.MAX_VALUE);
        Deque<Integer> dq = new ArrayDeque<Integer>(); // The trick to this problem is to use a deque instead of a queue
                                                        // If a vertex isn't a bridge, we can push it to the front of the deque instead

        dq.addLast(1);
        dist.set(1, 0);
        while(dq.size() > 0){
            int u = dq.pollFirst();
            for(int i = 0; i < adj.get(u).size(); i++){
                int v = adj.get(u).get(i);
                int new_dist = dist.get(u) + (is_single.get(u).get(i) ? 1 : 0);
                if(new_dist < dist.get(v)){
                    dist.set(v, new_dist);
                    if(new_dist == dist.get(u)){
                        dq.addFirst(v);
                    }else{
                        // It's fine if we push a vertex 2 times; the algorithm still takes O(n) time.
                        // Because we can show each vertex won't be pushed more than twice, 
                        // we don't need to do any processing to make sure we don't push the same vertices.
                        dq.addLast(v);
                    }
                }
            }
        }
        return dist.get(n);
    }

    public static void main(String[] args) throws Exception{
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(in.readLine());
        int n = Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
        ArrayList<ArrayList<Boolean>> is_single = new ArrayList<ArrayList<Boolean>>();
        while(adj.size() < n + 1) {
            adj.add(new ArrayList<Integer>());
            is_single.add(new ArrayList<Boolean>());
        }
        for(int i = 0; i < m; i++){
            st = new StringTokenizer(in.readLine());
            int u = Integer.parseInt(st.nextToken()), v = Integer.parseInt(st.nextToken()), z = Integer.parseInt(st.nextToken());
            adj.get(u).add(v);
            is_single.get(u).add(z == 1);
            adj.get(v).add(u);
            is_single.get(v).add(z == 1);
        }
        System.out.println(bridges(n, adj, is_single));
    }
}